### DISCUSSION TITLE
Generating an image with diagonal text

### DISCUSSION THREAD

**Discussion by madduck** [ImageMagick/ImageMagick#5036] [Help] [answered]
Hello,

Still stuck with ImageImagick 6.9.11, hope that isn't a problem.

I am trying to generate image of varying sizes that are basically just different texts diagonally, with the size of the text dynamically filling the available space.

I've tried multiple approaches, but they all fall short, and now I am kind of demotivated.

1. Apart from hard-coding the angle, which is a function of the image dimensions, this text is just the same size and does not scale dynamically. The `scale` transform only takes multiples, but I cannot know by what factor to scale the text because that's dynamic and dependent on image size and text length.
   ```
   convert -size 2550x3500 -density 300x300 xc:transparent \
     -gravity center -draw "fill grey rotate -60 text 0,0 'My text'"
   ```

2. So there is `label:`, and it can do the sizing:
   ```
   convert -size 2550x3500 -density 300x300 -background transparent \
     -gravity center -fill grey label:"My text" out.png
   ```
   but I cannot rotate that.

3. I can rotate the whole picture after the laben is written, but then the
resulting image is far larger than the desired dimensions:
   ```
   convert -size 2550x3500 -density 300x300 -background transparent \
     -gravity center -fill grey label:"This is a long text" -rotate -55
   ```

4. So I crop. But that's really confusing:
   ```
   convert -size 2550x3500 -density 300x300 -background transparent \
     -gravity center -fill grey label:"This is a long text" -rotate -55 \
     -crop 2550x3500+0+0 +repage out.png
   ```
   Despite `-gravity center`, the image seems to be cropped as if `-gravity
   southeast` was set, and I just cannot it to isolate the centre 2550×3500 pixels: ![image](https://user-images.githubusercontent.com/195073/162616267-8192ac75-36f5-4b8f-9ff2-2641e0067777.png)

Do you have any other ideas? Thank you!

**Comment by snibgo** [comment]
After "-rotate", you should often have "+repage".

I don't understand the problem. What are your given inputs, and what needs to be calculated within the command? What are the formulae for those calculations?

I assume the text is a given input. Is the "-pointsize" of the text also a given input? If it is, then the width and height of the text (before any rotation) will be calculated for you.

>Apart from hard-coding the angle, which is a function of the image dimensions, ...

So the angle is to be calculated. What from? The dimensions of the text? Or some final dimensions?

Are the final dimensions numbers that are given inputs, or should they be calculated? If calculated, what from?

You are using IM v6, which is rather old. Why? V7 is much better, with more ability to use formulae, when v6 had to use fixed numbers.

**Reply by madduck** [reply]
Hey @snibgo thanks for your time to reply!

To address your points:

1. I use `+repage` after `-crop`. That should be enough, right?
2. The angle can be calculated from the width and height of the target image, using the arc tangent, i.e. in this case `atan(3500/2550) ≅ 54°`. I am happy to do this outside of ImageMagick, so I don't have to use IM7 formulas.
3. I do *not* want to specify `-pointsize`. Instead, I want ImageMagick to use the largest possible font size such that the text fits across the diagonal.

I wish I could use IM7, but it's not in Debian, and I have not figured out how to install IM7.

I've made a bit of progress, but it's still not ideal. Consider this command:

```
convert -size 4330x -density 300x300 -background transparent -gravity center -fill grey label:"This is a long text" -rotate -54 -resize 2550x3500\! out.png
```

The 4330 is calculated using Pythagoras: `sqrt(2550² + 3500²)`. IM fills that whole width with the text label, scaling it appropriately, and automatically determines the height of the text.

This text is then rotated onto the diagonal of the image.

The final image is 3140×3936 pixels, which is the bounding box required to house the rotated 4330×x image. I then squash that into the right dimensions using `-resize` while ignoring the aspect ratio.

Better than nothing, but I am curious to hear how people would do this better.

**Comment by snibgo** [comment]
If I make many assumptions, a possible solution in Windows BAT is:
```
set WW=400

set HH=300

set TEXT=This is a text line

for /F "usebackq" %%L in (`%IM%convert xc: -format "%%[fx:hypot(%WW%,%HH%)]" info:`) do set HYPOT=%%L

for /F "usebackq" %%L in (`%IM%convert xc: -format "%%[fx:-atan2(%HH%,%WW%)*180/PI]" info:`) do set ANG=%%L

convert ^
  -size %HYPOT%x -background #88f label:"%TEXT%" ^
  -rotate %ANG% +repage -resize "%WW%x%HH%^!" ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/162627478-33d8a1ff-1c65-437f-991e-b38caf8f6e14.png)

**Reply by madduck** [reply]
Ha, we arrived at the same!

**Comment by madduck** [comment]
Here, without the distortion introduced by the brute-force resize:

```
convert -size 4330x -density 300x300 -background transparent -gravity center -fill grey label:"This is a long text" -rotate -54 +repage -resize 2480x3508 -extent 2480x3508 out.png
```

**Reply by madduck** [reply]
And here allowing for more than one line:

```
convert -size 4330x4330 -density 300x300 -background transparent -gravity center -fill grey label:"This is a very\nvery very long text" -trim -rotate -54 +repage -resize 2480x3508 -extent 2480x3508 out.png
```

**Comment by snibgo** [comment]
> 1. I use +repage after -crop. That should be enough, right?

No. A simple experiment shows this:
```
convert rose: -rotate 45 -crop 20x30+10+10 +repage x0.png

convert rose: -rotate 45 +repage -crop 20x30+10+10 +repage x1.png
```
The output images are different.

Yes, we came up with the similar solutions.

**Comment by yoya** [comment, marked as answer]
Will this work as expected?  (I just insert +repage after rotate)
I'm just guessing, but could it be that the origin is shifted by rotate and the +0+0 of crop is calculating the shifted-center offset.

```
% convert -size 2550x3500 -density 300x300 -background transparent \
  -gravity center -fill grey label:"This is a long text" -rotate -55 +repage \
  -crop 2550x3500+0+0 +repage out.png
% identify  out.png
out.png PNG 2550x3500 2550x3500+0+0 8-bit Grayscale Gray 98210B 0.000u 0:00.000
```

![out](https://user-images.githubusercontent.com/26040/162634209-aa26acb2-c25d-4a3b-a86b-f13159c3209d.png)

**Reply by madduck** [reply]
Nice, yeah, I am beginning to understand the use of `+repage`. Thank you!

**Comment by snibgo** [comment]
@yoya:
> Will this work as expected? (I just insert +repage after rotate)

That gives a different result.

When developing a command, seeing the image is much easier if you don't use a transparent background. Then we can clearly see the text doesn't extend to the image corners. Instead, the length of the text is the same as the image width (as we might expect).

EDIT: Also, if we don't resize, the text may be clipped at the image edges:
```
convert -size 2550x500 -density 300x300 -background #88f ^
  -gravity center -fill grey label:"This is a long text" -rotate -11 +repage ^
  -crop 2550x500+0+0 +repage out3.png
```

![out3](https://user-images.githubusercontent.com/33812368/162642517-d10f8cf4-f372-4c0f-b9ef-768147767243.png)

**Reply by yoya** [reply]
Thank you! I see.

The issue creator seemed to be concerned about -gravity, so I indicated its work-around solution.

If he wants the ability to leave the font bounding box by label: to apply after rotate as well, then he would need a script to calculation as snibgo suggested.

One could use -trim and -frame to get the minimum size to draw, but I feel that too naive.

```
% convert -size 2550x500 -density 300x300 -background "#88f" \
    -gravity center -fill grey label:"This is a long text" -rotate -11 \
    -trim -bordercolor "#88f" -border 32x64 +repage out-naive.png
```
![out-naive](https://user-images.githubusercontent.com/26040/162702046-500713e4-1a38-4f47-b635-a9574725615b.png)
